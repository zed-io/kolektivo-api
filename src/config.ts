import dotenv from 'dotenv'
import { logger } from './logger'

// Load environment variables from .env file
if (process.env.JEST_WORKER_ID) {
  dotenv.config({ path: './jest.env' })
} else {
  dotenv.config()
}

export function getFirebaseAdminCreds(admin: any) {
  if (DEPLOY_ENV === 'local') {
    try {
      const serviceAccount = require('../serviceAccountKey.json')
      return admin.credential.cert(serviceAccount)
    } catch (error) {
      logger.error({
        msg: 'Error: Could not initialize admin credentials. Is serviceAccountKey.json missing?',
        error,
      })
    }
  } else {
    try {
      return admin.credential.applicationDefault()
    } catch (error) {
      logger.error({
        msg: 'Error: Could not retrieve default app creds',
        error,
      })
    }
  }
}

export const DEPLOY_ENV = (process.env.DEPLOY_ENV as string).toLowerCase()
export const EXCHANGE_RATES_API = (
  process.env.EXCHANGE_RATES_API as string
).toLowerCase()
export const BLOCKSCOUT_API = (
  process.env.BLOCKSCOUT_API as string
).toLowerCase()
export const FIREBASE_PROJECT_ID = process.env.FIREBASE_PROJECT_ID
export const FIREBASE_DB = `https://${FIREBASE_PROJECT_ID}.firebaseio.com`
export const FAUCET_ADDRESS = (
  process.env.FAUCET_ADDRESS as string
).toLowerCase()
export const WEB3_PROVIDER_URL = process.env.WEB3_PROVIDER_URL
export const COINGECKO_API = (process.env.COINGECKO_API as string).toLowerCase()