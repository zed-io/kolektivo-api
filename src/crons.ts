import { ExchangeRateManager } from '@valora/exchanges'
import express, { Router } from 'express'
import { Knex } from 'knex'
import { logger } from './logger'
import PricesService from './prices/PricesService'
import {
  storeHistoricalPrices,
  updateCurrentPrices,
  updateHistoricalPrices,
} from './prices/PricesUpdater'

function addEndpoint(
  router: Router,
  path: string,
  errorType: string,
  asyncFn: () => Promise<void>,
) {
  router.get(path, async (req, res) => {
    // App Engine sets this header if and only if the request is from a cron.
    if (!req.headers['x-appengine-cron']) {
      logger.warn('Request does not contain header x-appengine-cron')
      res.status(401).send()
      return
    }

    try {
      await asyncFn()
      res.status(204).send()
    } catch (err) {
      logger.error({
        type: errorType,
        err,
      })
      res.status(500).send()
    }
  })
}

export function cronRouter({
  db,
  pricesService,
  exchangeRateManager,
}: {
  db: Knex
  pricesService: PricesService
  exchangeRateManager: ExchangeRateManager
}): Router {
  const router: Router = express.Router()

  addEndpoint(
    router,
    '/update-current-prices',
    'ERROR_UPDATING_CURRENT_PRICES',
    async () => {
      await updateCurrentPrices({ exchangeRateManager })
    },
  )

  addEndpoint(
    router,
    '/update-historical-prices',
    'ERROR_UPDATING_HISTORICAL_PRICES',
    async () => {
      await updateHistoricalPrices({ pricesService })
    },
  )

  addEndpoint(
    router,
    '/store-prices',
    'ERROR_UPDATING_HISTORICAL_PRICES',
    async () => {
      await storeHistoricalPrices({ db, exchangeRateManager })
    },
  )

  return router
}
