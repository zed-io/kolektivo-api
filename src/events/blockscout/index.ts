export * from './Any'
export * from './ContractCall'
export * from './EscrowReceived'
export * from './EscrowSent'
export * from './ExchangeCeloToToken'
export * from './ExchangeTokenToCelo'
export * from './TokenReceived'
export * from './TokenSent'
