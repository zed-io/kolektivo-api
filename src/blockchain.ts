import { DataSource } from 'apollo-datasource'
import { Address, createPublicClient, http } from 'viem'
import { celo } from 'viem/chains'
import { WEB3_PROVIDER_URL } from './config'
import { UserTokenBalance } from './types'
import tokenInfoCache from './helpers/TokenInfoCache'
import { erc20Abi } from './abis/ERC20'

export class BlockchainDataSource<TContext = any> extends DataSource {
  private client = createPublicClient({
    chain: celo,
    transport: http(WEB3_PROVIDER_URL),
  })

  constructor() {
    super()
  }

  async fetchUserBalances(address: string): Promise<UserTokenBalance[]> {
    const tokensInfo = tokenInfoCache.getTokensInfo()
    const tokensAddresses = tokensInfo.map((tokenInfo) =>
      tokenInfo.address.toLowerCase(),
    )

    const results = await this.client.multicall({
      contracts: tokensAddresses.map(
        (tokenAddress) =>
          ({
            address: tokenAddress as Address,
            abi: erc20Abi,
            functionName: 'balanceOf',
            args: [address as Address],
          } as const),
      ),
      allowFailure: false,
    })

    return results
      .map((result, index) => ({
        tokenAddress: tokensAddresses[index],
        balance: result.toString(),
        decimals: tokensInfo[index].decimals.toString(),
        symbol: tokensInfo[index].symbol,
      }))
      .filter((token) => token.balance !== '0')
  }
}
