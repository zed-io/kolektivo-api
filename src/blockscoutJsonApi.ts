import { RESTDataSource } from 'apollo-datasource-rest'
import { BLOCKSCOUT_API } from './config'
import { UserTokenBalance } from './types'

interface BlockscoutTokenBalance {
  balance: string
  contractAddress: string
  decimals: string
  name: string
  symbol: string
  type: string
}

export class BlockscoutJsonAPI extends RESTDataSource {
  constructor() {
    super()
    this.baseURL = `${BLOCKSCOUT_API}/api`
  }

  async fetchUserBalances(address: string): Promise<UserTokenBalance[]> {
    const response = await this.get(
      `?module=account&action=tokenlist&address=${address}`,
    )
    return response.result.map((row: BlockscoutTokenBalance) => ({
      tokenAddress: row.contractAddress,
      balance: row.balance,
      decimals: row.decimals,
      symbol: row.symbol,
    }))
  }
}
