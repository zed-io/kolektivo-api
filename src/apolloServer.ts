import { ApolloServer } from 'apollo-server-express'
import { BlockscoutAPI } from './blockscout'
import { BlockscoutJsonAPI } from './blockscoutJsonApi'
import CurrencyConversionAPI from './currencyConversion/CurrencyConversionAPI'
import OracleJsonAPI from './currencyConversion/OracleJsonAPI'
import { logger } from './logger'
import PricesService from './prices/PricesService'
import { resolvers } from './resolvers'
import typeDefs from './schema'

export interface DataSources {
  blockscoutAPI: BlockscoutAPI
  blockscoutJsonAPI: BlockscoutJsonAPI
  currencyConversionAPI: CurrencyConversionAPI
  pricesService: PricesService
  oracle: OracleJsonAPI
}

export function initApolloServer({
  currencyConversionAPI,
  pricesService,
  oracle,
}: {
  currencyConversionAPI: CurrencyConversionAPI
  pricesService: PricesService
  oracle: OracleJsonAPI
}) {
  return new ApolloServer({
    typeDefs,
    resolvers,
    introspection: true,
    dataSources: () => {
      return {
        blockscoutAPI: new BlockscoutAPI(),
        blockscoutJsonAPI: new BlockscoutJsonAPI(),
        currencyConversionAPI,
        pricesService,
        oracle,
      }
    },
    formatError: (error) => {
      logger.error({
        type: 'UNHANDLED_ERROR',
        error: error,
      })
      return error
    },
  })
}
