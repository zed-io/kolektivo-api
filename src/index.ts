import * as Sentry from '@sentry/node'
import * as Tracing from '@sentry/tracing'
import {
  configs as exchangesConfigs,
  createNewManager,
} from '@valora/exchanges'
import { loadSecret } from '@valora/secrets-loader'
import express from 'express'
import promBundle from 'express-prom-bundle'
import yargs from 'yargs'
import { initApolloServer } from './apolloServer'
import { cronRouter } from './crons'
import CurrencyConversionAPI from './currencyConversion/CurrencyConversionAPI'
import ExchangeRateAPI from './currencyConversion/ExchangeRateAPI'
import { initDatabase } from './database/db'
import knownAddressesCache from './helpers/KnownAddressesCache'
import tokenInfoCache from './helpers/TokenInfoCache'
import { logger } from './logger'
import PricesService from './prices/PricesService'

const metricsMiddleware = promBundle({ includeMethod: true, includePath: true })

const GRAPHQL_PATH: string = '/'

async function parseArgs() {
  //
  // Load secrets from Secrets Manager and inject into process.env.
  //
  const secretNames = process.env.SECRET_NAMES?.split(',') ?? []
  for (const secretName of secretNames) {
    Object.assign(process.env, await loadSecret(secretName))
  }

  const argv = yargs
    .env('')
    .option('port', {
      description: 'Port to listen on',
      type: 'number',
      default: 8080,
    })
    .option('exchanges-network-config', {
      description: 'Blockchain network config for exchanges',
      choices: Object.keys(exchangesConfigs),
      type: 'string',
      demandOption: true,
    })
    .option('exchange-rates-api-access-key', {
      description: 'API key for exchange-rates-api',
      type: 'string',
      demandOption: true,
    })
    .option('blockchain-db-host', {
      group: 'Blockchain DB:',
      description: 'Blockchain DB host',
      type: 'string',
      demandOption: true,
    })
    .option('blockchain-db-port', {
      group: 'Blockchain DB:',
      description: 'Blockchain DB host port',
      type: 'number',
      default: 5432,
    })
    .option('blockchain-db-database', {
      group: 'Blockchain DB:',
      description: 'Blockchain DB database',
      type: 'string',
      demandOption: true,
    })
    .option('blockchain-db-user', {
      group: 'Blockchain DB:',
      description: 'Blockchain DB user',
      type: 'string',
      demandOption: true,
    })
    .option('blockchain-db-pass', {
      group: 'Blockchain DB:',
      description: 'Blockchain DB pass',
      type: 'string',
      demandOption: true,
    })
    .option('sentry-dsn', {
      description: 'Sentry DSN',
      type: 'string',
    })
    .option('sentry-traces-sample-rate', {
      description: 'Sentry traces sample rate',
      type: 'number',
      default: 1.0,
    })
    .epilogue(
      'Always specify arguments as environment variables. Not all arguments are supported as CLI ones yet.',
    ).argv

  return argv
}

async function main() {
  const args = await parseArgs()
  const app = express()

  // The Sentry documentation recommends initializing as early in your app as
  // possible.
  const sentryEnabled = !!args['sentry-dsn']
  if (sentryEnabled) {
    Sentry.init({
      dsn: args['sentry-dsn'],
      integrations: [
        // enable HTTP calls tracing
        new Sentry.Integrations.Http({ tracing: true }),
        // enable Express.js middleware tracing
        new Tracing.Integrations.Express({ app }),
      ],
      tracesSampleRate: args['sentry-traces-sample-rate'],
    })

    app.use(Sentry.Handlers.requestHandler())
    app.use(Sentry.Handlers.tracingHandler())
  }

  const db = await initDatabase({
    client: 'pg',
    connection: {
      host: args['blockchain-db-host'],
      port: args['blockchain-db-port'],
      database: args['blockchain-db-database'],
      user: args['blockchain-db-user'],
      password: args['blockchain-db-pass'],
    },
  })

  const exchangeRateConfig = exchangesConfigs[args['exchanges-network-config']]
  const exchangeRateManager = createNewManager(exchangeRateConfig)

  knownAddressesCache.startListening()
  tokenInfoCache.startListening()

  const exchangeRateAPI = new ExchangeRateAPI({
    exchangeRatesAPIAccessKey: args['exchange-rates-api-access-key'],
  })
  const currencyConversionAPI = new CurrencyConversionAPI({ exchangeRateAPI })
  const pricesService = new PricesService(
    db,
    exchangeRateAPI,
    exchangeRateManager.cUSDTokenAddress,
  )

  app.use(metricsMiddleware)

  // What is this? lol
  app.get('/robots.txt', (_req, res) => {
    res.type('text/plain')
    res.send('User-agent: *\nDisallow: /')
  })

  app.head('/', (_req, res) => {
    // Preventing HEAD requests made by some browsers causing alerts
    // https://github.com/celo-org/celo-monorepo/issues/2189
    res.end()
  })

  app.use('/cron', cronRouter({ db, pricesService, exchangeRateManager }))

  const apolloServer = initApolloServer({
    currencyConversionAPI,
    pricesService,
  })
  await apolloServer.start()
  apolloServer.applyMiddleware({ app, path: GRAPHQL_PATH })

  if (sentryEnabled) {
    // The Sentry error handler must be before any other error middleware and
    // after all controllers
    app.use(Sentry.Handlers.errorHandler())
  }

  app.use(
    (
      error: Error,
      _req: express.Request,
      res: express.Response,
      _next: express.NextFunction,
    ) => {
      logger.error({ type: 'UNEXPECTED_ERROR', error })
      res.statusCode = 500
      if (sentryEnabled) {
        res.end((res as any).sentry + '\n')
      }
    },
  )

  app.listen(args.port, () => {
    logger.info(`Listening on port ${args.port}`)
    logger.info(`GraphQL path ${apolloServer.graphqlPath}`)
  })
}

main().catch((error) => {
  logger.error({
    type: 'STARTUP',
    error,
  })
  process.exit(1)
})
