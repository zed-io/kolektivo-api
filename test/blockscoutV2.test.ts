import { BlockscoutAPI } from '../src/blockscout'
import CurrencyConversionAPI from '../src/currencyConversion/CurrencyConversionAPI'
import mockTokenTxs, {
  TEST_DOLLAR_ADDRESS,
  TEST_GOLD_ADDRESS,
} from './mockTokenTxsV2'

const mockDataSourcePost = jest.fn(() => mockTokenTxs)

jest.mock('apollo-datasource-rest', () => {
  class MockRESTDataSource {
    baseUrl = ''
    post = mockDataSourcePost
  }

  return {
    RESTDataSource: MockRESTDataSource,
  }
})

jest.mock('../src/config.ts', () => {
  return {
    ...(jest.requireActual('../src/config.ts') as any),
    FAUCET_ADDRESS: '0x0000000000000000000000000000000000f40c37',
  }
})

jest.mock('../src/utils.ts', () => {
  const contractGetter = jest.fn()
  const tokenAddressMapping: { [key: string]: string } = {
    ['0x000000000000000000000000000000000000gold']: 'Celo Gold',
    ['0x0000000000000000000000000000000000dollar']: 'Celo Dollar',
  }
  contractGetter.mockReturnValue({
    tokenAddressMapping,
    GoldToken: '0x000000000000000000000000000000000000gold',
    StableToken: '0x0000000000000000000000000000000000dollar',
    Attestations: '0x0000000000000000000000000000000000a77357',
    Escrow: '0x0000000000000000000000000000000000a77327',
    Exchange: '0xf1235cb0d3703e7cc2473fb4e214fbc7a9ff77cc',
    ExchangeEUR: '0xd1235cb0d3703e7cc2473fb4e214fbc7a9ff77cc',
    ExchangeBRL: '0x4a3acb12178b40d8ca2b719cba6bcae0e8e31f4c',
    Governance: '0xa12a699c641cc875a7ca57495861c79c33d293b4',
    Reserve: '0x6a61e1e693c765cbab7e02a500665f2e13ee46df',
  })
  return {
    ...(jest.requireActual('../src/utils.ts') as any),
    getContractAddresses: contractGetter,
    getContractAddressesOrError: contractGetter,
  }
})

jest.mock('../src/helpers/KnownAddressesCache.ts', () => {
  return {
    startListening: {},
    getDisplayInfoFor: jest.fn().mockImplementation((address: string) => {
      switch (address) {
        case '0xf4314cb9046bece6aa54bb9533155434d0c76909':
          return { name: 'Test Name', imageUrl: 'Test Image' }
        case '0xa12a699c641cc875a7ca57495861c79c33d293b4':
          return { name: 'Test Only Name' }
        default:
          return {}
      }
    }),
  }
})

jest.mock('../src/helpers/TokenInfoCache.ts', () => ({
  getDecimalsForToken: (address: string) => {
    switch (address) {
      case TEST_GOLD_ADDRESS:
        return 18
      case TEST_DOLLAR_ADDRESS:
        return 12
      default:
        return 8
    }
  },
  tokenInfoBySymbol: (symbol: string) => {
    return {
      cUSD: {
        address: TEST_DOLLAR_ADDRESS.toLowerCase(),
      },
      CELO: {
        address: TEST_GOLD_ADDRESS.toLowerCase(),
      },
    }[symbol]
  },
  getTokensAddresses: () =>
    [TEST_GOLD_ADDRESS, TEST_DOLLAR_ADDRESS].map((address) =>
      address.toLowerCase(),
    ),
}))

// @ts-ignore
const mockCurrencyConversionAPI: CurrencyConversionAPI = {
  getFromMoneyAmount: jest.fn(),
}

const mockFetch = jest.fn()
jest.mock('../src/firebase.ts', () => ({
  fetchFromFirebase: (path: string) => mockFetch(path),
}))

describe('Blockscout', () => {
  let blockscoutAPI: BlockscoutAPI

  beforeEach(async () => {
    blockscoutAPI = new BlockscoutAPI()
    jest.clearAllMocks()
  })

  it('should return the same result if afterCursor is passed or not', async () => {
    const resultWithoutAfterCursor = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
    )

    const resultWithAfterCursor = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
    )

    expect(resultWithAfterCursor).toMatchObject(resultWithoutAfterCursor)
  })

  it('should get all transactions and label them properly when afterCursor is passed', async () => {
    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
    )

    // Reversing for convenience to match the order in mock data
    const transactions = result.transactions.reverse()
    expect(transactions).toMatchSnapshot()

    const pageInfo = result.pageInfo
    expect(pageInfo).toMatchSnapshot()
  })
})

describe('NFT events response according to wallet version', () => {
  let blockscoutAPI: BlockscoutAPI

  beforeEach(async () => {
    blockscoutAPI = new BlockscoutAPI()
    jest.clearAllMocks()
  })

  it('NFT events should not be included in the response if the version is less than 1.38.0', async () => {
    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      '1.4.0',
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(expect.not.stringContaining('NFT_SENT'))
    expect(resultString).toEqual(expect.not.stringContaining('NFT_RECEIVED'))
    expect(mockFetch).toHaveBeenCalledTimes(0)
  })

  it('NFT events should be included in the response if the version is equal or higher than 1.38.0', async () => {
    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      '1.38.0',
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(expect.stringContaining('NFT_SENT'))
    expect(resultString).toEqual(expect.stringContaining('NFT_RECEIVED'))
    expect(mockFetch).toHaveBeenCalledTimes(0)
  })

  it('fetches the app version from firebase when not set directly', async () => {
    mockFetch.mockImplementation((path: string) => {
      return {
        appVersion: '1.38.0',
      }
    })

    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      undefined,
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(expect.stringContaining('NFT_SENT'))
    expect(resultString).toEqual(expect.stringContaining('NFT_RECEIVED'))
    expect(mockFetch).toHaveBeenCalledTimes(1)
  })
})

describe('Swap events response according to wallet version', () => {
  let blockscoutAPI: BlockscoutAPI

  beforeEach(async () => {
    blockscoutAPI = new BlockscoutAPI()
    jest.clearAllMocks()
  })

  it('Swap events should not be included in the response if the version is less than 1.39.0', async () => {
    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      '1.4.0',
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(
      expect.not.stringContaining('SWAP_TRANSACTION'),
    )
    expect(mockFetch).toHaveBeenCalledTimes(0)
  })

  it('Swap events should be included in the response if the version is equal or higher than 1.39.0', async () => {
    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      '1.39.0',
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(expect.stringContaining('SWAP_TRANSACTION'))
    expect(mockFetch).toHaveBeenCalledTimes(0)
  })

  it('fetches the app version from firebase when not set directly', async () => {
    mockFetch.mockImplementation((path: string) => {
      return {
        appVersion: '1.39.0',
      }
    })

    const result = await blockscoutAPI.getTokenTransactionsV2(
      '0x0000000000000000000000000000000000007E57',
      'afterCursor',
      undefined,
    )

    const resultString = JSON.stringify(result)

    expect(resultString).toEqual(expect.stringContaining('SWAP_TRANSACTION'))
    expect(mockFetch).toHaveBeenCalledTimes(1)
  })
})
