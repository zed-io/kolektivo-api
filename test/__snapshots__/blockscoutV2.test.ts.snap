// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockscout should get all transactions and label them properly when afterCursor is passed 1`] = `
[
  {
    "block": "90608",
    "fees": [
      {
        "amount": {
          "timestamp": 1566345797000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566345797000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.010091957572465625",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "inAmount": {
      "timestamp": 1566345797000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "10000000",
    },
    "outAmount": {
      "timestamp": 1566345797000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "1",
    },
    "timestamp": 1566345797000,
    "transactionHash": "0xba620de2d812f299d987155eb5dca7abcfeaf154f5cfd99cb1773452a7df3d7a",
    "type": "EXCHANGE",
  },
  {
    "block": "90637",
    "fees": [
      {
        "amount": {
          "timestamp": 1566345926000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566345926000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.01102543453093182",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "inAmount": {
      "timestamp": 1566345926000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "1",
    },
    "outAmount": {
      "timestamp": 1566345926000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "10000000",
    },
    "timestamp": 1566345926000,
    "transactionHash": "0x961403536006f9c120c23900f94da59dbf43edf10eb3569b448665483bab77b2",
    "type": "EXCHANGE",
  },
  {
    "account": "0x8b7649116f169d2d2aebb6ea1a77f0baf31f2811",
    "address": "0x8b7649116f169d2d2aebb6ea1a77f0baf31f2811",
    "amount": {
      "timestamp": 1566346276000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-150000",
    },
    "block": "90719",
    "fees": [
      {
        "amount": {
          "timestamp": 1566346276000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.01",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566346276000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.0056589",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1566346276000,
    "transactionHash": "0x21dd2c18ae6c80d61ffbddaa073f7cde7bbfe9436fdf5059b506f1686326a2fb",
    "type": "SENT",
  },
  {
    "account": "0xa12a699c641cc875a7ca57495861c79c33d293b4",
    "address": "0xa12a699c641cc875a7ca57495861c79c33d293b4",
    "amount": {
      "timestamp": 1566346336000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-150000",
    },
    "block": "90791",
    "fees": [
      {
        "amount": {
          "timestamp": 1566346336000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.01",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566346336000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.0056589",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": "Test Only Name",
    },
    "timestamp": 1566346336000,
    "transactionHash": "0x21dd2c18ae6c80d61ffbddaa073f7cde7bbfe9436fdf5059b506f1686326a2ff",
    "type": "SENT",
  },
  {
    "account": "0x6a0edf42f5e618bee697e7718fa05efb1ea5d11c",
    "address": "0x6a0edf42f5e618bee697e7718fa05efb1ea5d11c",
    "amount": {
      "timestamp": 1566346396000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-150000",
    },
    "block": "90792",
    "fees": [
      {
        "amount": {
          "timestamp": 1566346396000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.01",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566346396000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.0056589",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1566346396000,
    "transactionHash": "0x21dd2c18ae6c80d61ffbddaa073f7cde7bbfe9436fdf5059b506f1686326afff",
    "type": "SENT",
  },
  {
    "account": "0xf4314cb9046bece6aa54bb9533155434d0c76910",
    "address": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "amount": {
      "timestamp": 1566479946000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "10000000",
    },
    "block": "117453",
    "fees": [],
    "metadata": {
      "comment": "",
      "image": "Test Image",
      "title": "Test Name",
    },
    "timestamp": 1566479946000,
    "transactionHash": "0xe70bf600802bae7a0d42d89d54b8cdb977a8c5a34a239ec73597c7abcab74536",
    "type": "RECEIVED",
  },
  {
    "account": "0xf4314cb9046bece6aa54bb9533155434d0c76910",
    "address": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "amount": {
      "timestamp": 1566481000000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "-1",
    },
    "block": "117451",
    "fees": [
      {
        "amount": {
          "timestamp": 1566481000000,
          "tokenAddress": "0x000000000000000000000000000000000000gold",
          "value": "0.00023713",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": "Test Image",
      "title": "Test Name",
    },
    "timestamp": 1566481000000,
    "transactionHash": "0xc6689ed516e8b114e875d682bbf7ba318ea16841711d97ce473f20da289435be",
    "type": "SENT",
  },
  {
    "account": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "address": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "amount": {
      "timestamp": 1566481060000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "-1",
    },
    "block": "117451",
    "fees": [
      {
        "amount": {
          "timestamp": 1566481060000,
          "tokenAddress": "0x000000000000000000000000000000000000gold",
          "value": "0.00000001",
        },
        "type": "GATEWAY_FEE",
      },
      {
        "amount": {
          "timestamp": 1566481060000,
          "tokenAddress": "0x000000000000000000000000000000000000gold",
          "value": "0.00023713",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": "Test Image",
      "title": "Test Name",
    },
    "timestamp": 1566481060000,
    "transactionHash": "0xc6689ed516e8b114e875d682bbf7ba318ea16841711d97ce473f20da289435bd",
    "type": "SENT",
  },
  {
    "account": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "address": "0xf4314cb9046bece6aa54bb9533155434d0c76909",
    "amount": {
      "timestamp": 1566482000000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "10",
    },
    "block": "117451",
    "fees": [],
    "metadata": {
      "comment": "",
      "image": "Test Image",
      "title": "Test Name",
    },
    "timestamp": 1566482000000,
    "transactionHash": "0xe8fe81f455eb34b672a8d8dd091472f1ae8d4d204817f0bcbb7a13486b9b5605",
    "type": "RECEIVED",
  },
  {
    "account": "0x0000000000000000000000000000000000f40c37",
    "address": "0x0000000000000000000000000000000000f40c37",
    "amount": {
      "timestamp": 1566483000000,
      "tokenAddress": "0x000000000000000000000000000000000000golD",
      "value": "5",
    },
    "block": "117451",
    "fees": [],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1566483000000,
    "transactionHash": "0xf6856169eb7bf78211babc312028cddf3dad2761799428ab6e4fcf297a27fe09",
    "type": "RECEIVED",
  },
  {
    "account": "0x0000000000000000000000000000000000a77327",
    "address": "0x0000000000000000000000000000000000a77327",
    "amount": {
      "timestamp": 1566483998000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-118829.058457955309",
    },
    "block": "117451",
    "fees": [
      {
        "amount": {
          "timestamp": 1566483998000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.0005911869963022",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1566483998000,
    "transactionHash": "0xf0592e026656f84cc17672fb08f5723deb8426787c2865aa763e859d10e85234",
    "type": "INVITE_SENT",
  },
  {
    "account": "0x0000000000000000000000000000000000a77327",
    "address": "0x0000000000000000000000000000000000a77327",
    "amount": {
      "timestamp": 1618593019000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "20000000",
    },
    "block": "6203333",
    "fees": [],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1618593019000,
    "transactionHash": "0x8323ad9ea0961221e146aa086b9cda858acbf7a4058a23061c9e0d52a0e2d9b1",
    "type": "INVITE_RECEIVED",
  },
  {
    "account": "0x0000000000000000000000000000000000a77357",
    "address": "0x0000000000000000000000000000000000a77357",
    "amount": {
      "timestamp": 1618593079000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-200000",
    },
    "block": "117451",
    "fees": [
      {
        "amount": {
          "timestamp": 1618593079000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.00795255",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1618593079000,
    "transactionHash": "0xcc2120e5d050fd68284dc01f6464b2ed8f7358ca80fccb20967af28eb7d79160",
    "type": "SENT",
  },
  {
    "account": "0x0000000000000000000000000000000000007E57",
    "address": "0x0000000000000000000000000000000000007E57",
    "amount": {
      "timestamp": 1621185010000,
      "tokenAddress": "0x0000000000000000000000000000000000dollaR",
      "value": "-1000000",
    },
    "block": "1487877",
    "fees": [
      {
        "amount": {
          "timestamp": 1621185010000,
          "tokenAddress": "0x0000000000000000000000000000000000dollar",
          "value": "0.000051519",
        },
        "type": "SECURITY_FEE",
      },
    ],
    "metadata": {
      "comment": "",
      "image": undefined,
      "title": undefined,
    },
    "timestamp": 1621185010000,
    "transactionHash": "0x34e6e74bc01c7112817e669a8057ae7f4c1ed49d8de824bea8ecbdb945b41345",
    "type": "SENT",
  },
]
`;

exports[`Blockscout should get all transactions and label them properly when afterCursor is passed 2`] = `
{
  "endCursor": "YXJyYXljb25uZWN0aW9uOjI",
  "hasNextPage": true,
  "hasPreviousPage": false,
  "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
}
`;
